### PR TITLE
Made some of the public DataServiCeContext methods virtual to support mocking

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes issue #1250*

### Description

*Made some of the DataServiceContext public methods virtual to support mocking*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Comment by xuzhg** [comment]
>         public bool Detach(object entity)

how about this method ?
 or others?

why don't make them as virtual?

---
Refers to: src/Microsoft.OData.Client/DataServiceContext.cs:2229 in 5234ce9. [](commit_id = 5234ce9252f2e13e1b24b96553e7084596468153, deletion_comment = False)

File: src/Microsoft.OData.Client/DataServiceContext.cs [hunk 1]
```diff
@@ -2749,7 +2749,7 @@ internal BuildingRequestEventArgs CreateRequestArgsAndFireBuildingRequest(string
         /// <param name="languageDependentNamespace">Namespace of the type, can be different in VB than the fullNamespace.</param>
         /// <returns>Type that the name resolved to or null if none found.</returns>
         /// <remarks>Function was added for Portable Lib support to handle the differences in accessing the assembly of the context.</remarks>
-        protected internal Type DefaultResolveType(string typeName, string fullNamespace, string languageDependentNamespace)
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/DataServiceContext.cs]: >protected internal [](start = 8, length = 18)

I don't think to change this "internal" as public? What's the reason why we should do like this.

**Comment by xuzhg** [comment]
>         public IEnumerable EndExecute(IAsyncResult asyncResult)

what about this method?

---
Refers to: src/Microsoft.OData.Client/DataServiceQuery.cs:78 in 5234ce9. [](commit_id = 5234ce9252f2e13e1b24b96553e7084596468153, deletion_comment = False)

**Comment by xuzhg** [comment]
>         public bool IsComposable { get; private set; }

how about other public properties?

---
Refers to: src/Microsoft.OData.Client/DataServiceQueryOfT.cs:125 in 5234ce9. [](commit_id = 5234ce9252f2e13e1b24b96553e7084596468153, deletion_comment = False)

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj [hunk 2]
```diff
@@ -15,6 +15,7 @@
     <PackageReference Include="FluentAssertions" Version="4.19.2" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.0.0" />
     <PackageReference Include="Microsoft.NETCore.Portable.Compatibility" Version="1.0.1" />
+    <PackageReference Include="Moq" Version="4.6.38-alpha" />
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: >"4.6.38-alpha" /> [](start = 44, length = 17)

why do we use the "alpha" version? please use the latest RTM version

File: test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+using Moq;
+using Xunit;
+
+namespace Microsoft.OData.Client.Tests.ALinq
+{
+    /// <summary>
+    /// How to mock the DataServiceContext class methods using moq framework example.
+    /// </summary>
+
+    public class DataServiceContextMethodMockSampleTest
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs]: >DataServiceContextMethodMockSampleTest [](start = 17, length = 38)

I don't think the test class is meaningful.

What we are testing is that we can test some functionalities using mock, not test mock itself.

For example, we can test some method is actually called when we call a certain method. And we don't care about that called method returns, we can use mock to mock that called method.

File: tools/CustomMSBuild/Versioning.props [hunk 4]
```diff
@@ -18,7 +18,7 @@
     -->
   <PropertyGroup>
     <VersionStartYear Condition="'$(VersionStartYear)' == ''">2017</VersionStartYear>
-    <VersionDateCode>$([System.Convert]::ToInt16('$([MSBuild]::Add(1, $([MSBuild]::Subtract($([System.DateTime]::Now.Year), $(VersionStartYear)))))$([System.DateTime]::Now.ToString("MMdd"))'))</VersionDateCode>
+    <VersionDateCode>$([System.Convert]::ToInt32('$([MSBuild]::Add(1, $([MSBuild]::Subtract($([System.DateTime]::Now.Year), $(VersionStartYear)))))$([System.DateTime]::Now.ToString("MMdd"))'))</VersionDateCode>
```
Comment (xuzhg) [inline tools/CustomMSBuild/Versioning.props]: >ToInt32 [](start = 41, length = 7)

This problem is fixed. Please rebase your PR to the latest master branch

File: test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            DoSomething doSomething = new DoSomething() ;
+            string result = doSomething.SaveUser(mockObject.Object);
+            Assert.Equal(str, result);
+        }
+    }
+
+    /// <summary>
+    /// A sample implementation class with methods calling the DataserviceContext.cs methods
+    /// </summary>
+    public class DoSomething
```
Comment (odero) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs]: While I realize this is just a test class I'd avoid namings like `DoSomething` in publicly consumed code

Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs]: fixed and well noted

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj (hunk 2 above)
Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.Netcore.csproj]: It is the only version that is compatible with the .Netcore project.

File: test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            SaveUserDetails saveUserDetails = new SaveUserDetails() ;
+            string result = saveUserDetails.SaveUser(mockObject.Object);
+            Assert.Equal(str, result);
+        }
+    }
+
+    /// <summary>
+    /// A sample implementation class with methods calling the DataserviceContext.cs methods
+    /// </summary>
+    public class SaveUserDetails
```
Comment (wachugamaina) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs]: Class Names, variable names and property names ideally should be Nouns, only functions, actions and methods should be verbs.

**Pull request closed without merging**